server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java

```diff
@@ -33,58 +39,78 @@
  *
  * @opensearch.experimental
  */
-public class OpenSearchOnHeapCache<K, V> implements ICache<K, V>, RemovalListener<K, V> {
+public class OpenSearchOnHeapCache<K, V> implements ICache<K, V>, RemovalListener<ICacheKey<K>, V> {
 
-    private final Cache<K, V> cache;
-    private final RemovalListener<K, V> removalListener;
+    private final Cache<ICacheKey<K>, V> cache;
+    private CacheStats stats;
+    private final RemovalListener<ICacheKey<K>, V> removalListener;
+    private final List<String> dimensionNames;
+    public static final String TIER_DIMENSION_VALUE = "on_heap";
 
     public OpenSearchOnHeapCache(Builder<K, V> builder) {
-        CacheBuilder<K, V> cacheBuilder = CacheBuilder.<K, V>builder()
+        CacheBuilder<ICacheKey<K>, V> cacheBuilder = CacheBuilder.<ICacheKey<K>, V>builder()
             .setMaximumWeight(builder.getMaxWeightInBytes())
             .weigher(builder.getWeigher())
             .removalListener(this);
         if (builder.getExpireAfterAcess() != null) {
             cacheBuilder.setExpireAfterAccess(builder.getExpireAfterAcess());
         }
         cache = cacheBuilder.build();
+        this.dimensionNames = Objects.requireNonNull(builder.dimensionNames, "Dimension names can't be null");
+        this.stats = new MultiDimensionCacheStats(dimensionNames, TIER_DIMENSION_VALUE);
         this.removalListener = builder.getRemovalListener();
     }
 
     @Override
-    public V get(K key) {
+    public V get(ICacheKey<K> key) {
         V value = cache.get(key);
+        if (value != null) {
+            stats.incrementHitsByDimensions(key.dimensions);
```

Comment:
Maybe we should keep a common class that handles the incrementing/storing logic from MultiDimensionCacheStats, something like InternalCacheStats, which would expose the increment/decrement methods. Otherwise we'd have to reimplement basically the same thing with maps for each tier, and handling multiple dimensions would be tricky that way. 

But only the cache would have access to its private InternalCacheStats instance, so other classes changing it wouldn't be an issue. Calling `cache.stats()` would return something like `new MultiDimensionStats(internalStats)`, which would not expose mutating methods. What do you think about this?